File: backend/tests/integration/presets.test.ts
/**
 * Integration tests for Preset CRUD API Endpoints (TASK-4602).
 *
 * Tests:
 *   GET /api/presets              -> [] for a new user, array of preset objects otherwise
 *   POST /api/presets             -> 201 with created preset record
 *   POST /api/presets (limit hit) -> 422 with { error: 'PRESET_LIMIT_EXCEEDED' }
 *   PUT /api/presets/:id          -> 200 with updated preset
 *   PUT /api/presets/:id (other)  -> 403/404 for another user's preset
 *   DELETE /api/presets/:id       -> 204 with empty body
 *   Zod validation                -> 422 when name is missing or exceeds 100 chars
 *
 * Mocks the presetService so no real database access is required.
 */

import request from 'supertest';
import { jest } from '@jest/globals';

// Mock the preset service BEFORE importing the app so that the Express routes
// use the mock instead of the real service.
jest.mock('../../src/services/preset.service', () => ({
  presetService: {
    getPresets: jest.fn(),
    createPreset: jest.fn(),
    updatePreset: jest.fn(),
    deletePreset: jest.fn(),
  },
  PresetLimitExceededError: class PresetLimitExceededError extends Error {
    statusCode = 422;
    constructor() {
      super('You may not have more than 5 presets');
      this.name = 'PresetLimitExceededError';
    }
  },
}));

import app from '../../src/app';

// ---------------------------------------------------------------------------
// Test fixtures and mock references
// ---------------------------------------------------------------------------

const TEST_USER_ID = 'user-123';
const OTHER_USER_ID = 'other-user-456';

// Header providing authenticated userId for test requests
const authHeader = { 'x-user-id': TEST_USER_ID };

const testPreset = {
  id: 'preset-001',
  userId: TEST_USER_ID,
  name: 'Test Preset',
  branchName: 'main',
  commitSha: null,
  envVars: { NODE_ENV: 'test' },
  featureFlags: { featureA: true },
  createdAt: new Date('2024-01-01T00:00:00Z').toISOString(),
  updatedAt: new Date('2024-01-01T00:00:00Z').toISOString(),
};

// Get typed references to the mock functions
const { presetService: mockService, PresetLimitExceededError } =
  jest.requireMock('../../src/services/preset.service') as {
    presetService: {
      getPresets: jest.Mock;
      createPreset: jest.Mock;
      updatePreset: jest.Mock;
      deletePreset: jest.Mock;
    };
    PresetLimitExceededError: new () => Error & { statusCode: number };
  };

// ---------------------------------------------------------------------------
// GET /api/presets
// ---------------------------------------------------------------------------

describe('GET /api/presets', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockService.getPresets.mockResolvedValue([]);
  });

  it('returns 200 with an empty array for a new user', async () => {
    const res = await request(app)
      .get('/api/presets')
      .set(authHeader);

    expect(res.status).toBe(200);
    expect(Array.isArray(res.body)).toBe(true);
    expect(res.body).toHaveLength(0);
  });

  it('returns 200 with an array of preset objects when user has presets', async () => {
    mockService.getPresets.mockResolvedValue([testPreset]);

    const res = await request(app)
      .get('/api/presets')
      .set(authHeader);

    expect(res.status).toBe(200);
    expect(Array.isArray(res.body)).toBe(true);
    expect(res.body.length).toBeGreaterThan(0);
  });

  it('each preset object has id, name, userId, envVars, featureFlags fields', async () => {
    mockService.getPresets.mockResolvedValue([testPreset]);

    const res = await request(app)
      .get('/api/presets')
      .set(authHeader);

    expect(res.status).toBe(200);
    for (const preset of res.body) {
      expect(preset).toHaveProperty('id');
      expect(preset).toHaveProperty('name');
      expect(preset).toHaveProperty('userId');
      expect(preset).toHaveProperty('envVars');
      expect(preset).toHaveProperty('featureFlags');
    }
  });

  it('passes the authenticated userId to getPresets', async () => {
    await request(app)
      .get('/api/presets')
      .set(authHeader);

    expect(mockService.getPresets).toHaveBeenCalledWith(TEST_USER_ID);
  });
});

// ---------------------------------------------------------------------------
// POST /api/presets
// ---------------------------------------------------------------------------

describe('POST /api/presets', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockService.createPreset.mockResolvedValue(testPreset);
  });

  it('creates a preset and returns 201 with the preset record', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'Test Preset', branchName: 'main', envVars: {}, featureFlags: {} });

    expect(res.status).toBe(201);
    expect(res.body).toHaveProperty('id');
    expect(res.body).toHaveProperty('name');
  });

  it('returned preset record contains expected fields', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'Test Preset' });

    expect(res.status).toBe(201);
    expect(res.body).toHaveProperty('id');
    expect(res.body).toHaveProperty('userId');
    expect(res.body).toHaveProperty('name');
    expect(res.body).toHaveProperty('envVars');
    expect(res.body).toHaveProperty('featureFlags');
  });

  it('returns 422 with { error: PRESET_LIMIT_EXCEEDED } when user already has 5 presets', async () => {
    mockService.createPreset.mockRejectedValueOnce(new PresetLimitExceededError());

    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'New Preset' });

    expect(res.status).toBe(422);
    expect(res.body).toHaveProperty('error');
    expect(res.body.error).toBe('PRESET_LIMIT_EXCEEDED');
  });

  it('response body contains PRESET_LIMIT_EXCEEDED key (not a generic message) when limit hit', async () => {
    mockService.createPreset.mockRejectedValueOnce(new PresetLimitExceededError());

    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'Sixth Preset' });

    expect(res.status).toBe(422);
    expect(res.body.error).toBe('PRESET_LIMIT_EXCEEDED');
  });

  // ---- Zod validation ----

  it('returns 422 when name field is missing (zod validation)', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ branchName: 'main', envVars: {} });

    expect(res.status).toBe(422);
  });

  it('returns 422 when name is an empty string (min 1 char zod validation)', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: '', branchName: 'main' });

    expect(res.status).toBe(422);
  });

  it('returns 422 when name exceeds 100 characters (zod max length validation)', async () => {
    const longName = 'a'.repeat(101);

    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: longName });

    expect(res.status).toBe(422);
  });

  it('accepts a name of exactly 100 characters (boundary - zod max)', async () => {
    const maxName = 'a'.repeat(100);

    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: maxName });

    expect(res.status).toBe(201);
  });

  it('accepts optional branchName and commitSha fields', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'Preset with branch', branchName: 'feature/x', commitSha: 'abc123' });

    expect(res.status).toBe(201);
  });

  it('accepts optional envVars as a key-value record', async () => {
    const res = await request(app)
      .post('/api/presets')
      .set(authHeader)
      .send({ name: 'Preset', envVars: { KEY: 'VALUE', ANOTHER: 'val' } });

    expect(res.status).toBe(201);
  });
});

// ---------------------------------------------------------------------------
// PUT /api/presets/:id
// ---------------------------------------------------------------------------

describe('PUT /api/presets/:id', () => {
  const updatedPreset = { ...testPreset, name: 'Updated Preset', envVars: { KEY: 'value' } };

  beforeEach(() => {
    jest.clearAllMocks();
    mockService.updatePreset.mockResolvedValue(updatedPreset);
  });

  it('updates a preset and returns 200 with the updated record', async () => {
    const res = await request(app)
      .put('/api/presets/preset-001')
      .set(authHeader)
      .send({ name: 'Updated Preset', envVars: { KEY: 'value' } });

    expect(res.status).toBe(200);
    expect(res.body).toHaveProperty('name');
  });

  it('returns the updated name in the response body', async () => {
    const res = await request(app)
      .put('/api/presets/preset-001')
      .set(authHeader)
      .send({ name: 'Updated Preset' });

    expect(res.status).toBe(200);
    expect(res.body.name).toBe('Updated Preset');
  });

  it('returns the updated envVars in the response body', async () => {
    const res = await request(app)
      .put('/api/presets/preset-001')
      .set(authHeader)
      .send({ envVars: { KEY: 'value' } });

    expect(res.status).toBe(200);
    expect(res.body).toHaveProperty('envVars');
  });

  it("returns 403 or 404 when trying to update another user's preset", async () => {
    mockService.updatePreset.mockRejectedValueOnce(
      Object.assign(new Error('Preset not found'), { statusCode: 404 })
    );

    const res = await request(app)
      .put('/api/presets/other-users-preset')
      .set({ 'x-user-id': OTHER_USER_ID })
      .send({ name: 'Hacked Name' });

    expect([403, 404]).toContain(res.status);
  });

  it('returns 403 or 404 for a nonexistent preset id', async () => {
    mockService.updatePreset.mockRejectedValueOnce(
      Object.assign(new Error('Preset not found'), { statusCode: 404 })
    );

    const res = await request(app)
      .put('/api/presets/nonexistent-id')
      .set(authHeader)
      .send({ name: 'New Name' });

    expect([403, 404]).toContain(res.status);
  });

  // ---- Zod validation on update ----

  it('returns 422 when name exceeds 100 characters on update (zod validation)', async () => {
    const longName = 'a'.repeat(101);

    const res = await request(app)
      .put('/api/presets/preset-001')
      .set(authHeader)
      .send({ name: longName });

    expect(res.status).toBe(422);
  });
});

// ---------------------------------------------------------------------------
// DELETE /api/presets/:id
// ---------------------------------------------------------------------------

describe('DELETE /api/presets/:id', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockService.deletePreset.mockResolvedValue(undefined);
  });

  it('deletes a preset and returns 204', async () => {
    const res = await request(app)
      .delete('/api/presets/preset-001')
      .set(authHeader);

    expect(res.status).toBe(204);
  });

  it('returns an empty body on successful deletion', async () => {
    const res = await request(app)
      .delete('/api/presets/preset-001')
      .set(authHeader);

    expect(res.status).toBe(204);
    expect(res.body).toEqual({});
  });

  it('passes the preset id and userId to deletePreset', async () => {
    await request(app)
      .delete('/api/presets/preset-001')
      .set(authHeader);

    expect(mockService.deletePreset).toHaveBeenCalledWith('preset-001', TEST_USER_ID);
  });

  it("returns 403 or 404 when trying to delete another user's preset", async () => {
    mockService.deletePreset.mockRejectedValueOnce(
      Object.assign(new Error('Preset not found'), { statusCode: 404 })
    );

    const res = await request(app)
      .delete('/api/presets/other-users-preset')
      .set({ 'x-user-id': OTHER_USER_ID });

    expect([403, 404]).toContain(res.status);
  });

  it('returns 403 or 404 for a nonexistent preset id', async () => {
    mockService.deletePreset.mockRejectedValueOnce(
      Object.assign(new Error('Preset not found'), { statusCode: 404 })
    );

    const res = await request(app)
      .delete('/api/presets/nonexistent-id')
      .set(authHeader);

    expect([403, 404]).toContain(res.status);
  });
});

File: backend/tests/integration/presets_test.go
// Package integration_test validates the Preset CRUD API files for TASK-4602.
// These are file-system tests that verify the routes, controller, and integration
// test TypeScript files exist and contain all required definitions.
//
// Tests for routes/controller SHOULD FAIL until the implementation is created.
package integration_test

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// ---- Helpers ----

func readPresetsRoutesTS(t *testing.T) string {
	t.Helper()
	path := filepath.Join(repoRoot(t), "backend", "src", "routes", "presets.routes.ts")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read presets.routes.ts: %v", err)
	}
	return string(data)
}

func readPresetsControllerTS(t *testing.T) string {
	t.Helper()
	path := filepath.Join(repoRoot(t), "backend", "src", "controllers", "presets.controller.ts")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read presets.controller.ts: %v", err)
	}
	return string(data)
}

func readPresetsIntegrationTestTS(t *testing.T) string {
	t.Helper()
	path := filepath.Join(repoRoot(t), "backend", "tests", "integration", "presets.test.ts")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read presets.test.ts: %v", err)
	}
	return string(data)
}

// ---- File existence tests ----

func TestPresetsRoutesFileExists(t *testing.T) {
	path := filepath.Join(repoRoot(t), "backend", "src", "routes", "presets.routes.ts")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Fatalf("presets.routes.ts not found at %s", path)
	}
}

func TestPresetsControllerFileExists(t *testing.T) {
	path := filepath.Join(repoRoot(t), "backend", "src", "controllers", "presets.controller.ts")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Fatalf("presets.controller.ts not found at %s", path)
	}
}

func TestPresetsIntegrationTestFileExists(t *testing.T) {
	path := filepath.Join(repoRoot(t), "backend", "tests", "integration", "presets.test.ts")
	if _, err := os.Stat(path); os.IsNotExist(err) {
		t.Fatalf("presets.test.ts not found at %s", path)
	}
}

// ---- presets.routes.ts content ----

func TestPresetsRoutesDefinesGetRoute(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, ".get(") && !strings.Contains(lower, "router.get") {
		t.Error("presets.routes.ts must define a GET route for listing presets")
	}
}

func TestPresetsRoutesDefinesPostRoute(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, ".post(") && !strings.Contains(lower, "router.post") {
		t.Error("presets.routes.ts must define a POST route for creating presets")
	}
}

func TestPresetsRoutesDefinesPutRoute(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, ".put(") && !strings.Contains(lower, "router.put") {
		t.Error("presets.routes.ts must define a PUT route for updating presets")
	}
}

func TestPresetsRoutesDefinesDeleteRoute(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, ".delete(") && !strings.Contains(lower, "router.delete") {
		t.Error("presets.routes.ts must define a DELETE route for removing presets")
	}
}

func TestPresetsRoutesUsesIdParam(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	if !strings.Contains(routes, ":id") {
		t.Error("presets.routes.ts must include :id parameter for single-preset routes (PUT, DELETE)")
	}
}

func TestPresetsRoutesImportsRouter(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	if !strings.Contains(routes, "Router") {
		t.Error("presets.routes.ts must use Express Router")
	}
}

func TestPresetsRoutesImportsController(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, "controller") {
		t.Error("presets.routes.ts must import from the presets controller")
	}
}

func TestPresetsRoutesScopedToApiPresets(t *testing.T) {
	routes := readPresetsRoutesTS(t)
	lower := strings.ToLower(routes)
	if !strings.Contains(lower, "presets") {
		t.Error("presets.routes.ts must be scoped under /api/presets path")
	}
}

// ---- presets.controller.ts content ----

func TestPresetsControllerExportsGetPresets(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "getPresets") {
		t.Error("presets.controller.ts must export a getPresets handler")
	}
}

func TestPresetsControllerExportsCreatePreset(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "createPreset") {
		t.Error("presets.controller.ts must export a createPreset handler")
	}
}

func TestPresetsControllerExportsUpdatePreset(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "updatePreset") {
		t.Error("presets.controller.ts must export an updatePreset handler")
	}
}

func TestPresetsControllerExportsDeletePreset(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "deletePreset") {
		t.Error("presets.controller.ts must export a deletePreset handler")
	}
}

func TestPresetsControllerExtractsUserId(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "userId") {
		t.Error("presets.controller.ts must extract userId from the request context (auth middleware)")
	}
}

func TestPresetsControllerReturns201OnCreate(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "201") {
		t.Error("presets.controller.ts must return 201 status on successful preset creation")
	}
}

func TestPresetsControllerReturns204OnDelete(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "204") {
		t.Error("presets.controller.ts must return 204 status on successful preset deletion")
	}
}

func TestPresetsControllerHandlesPresetLimitExceededError(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "422") && !strings.Contains(ctrl, "statusCode") && !strings.Contains(ctrl, "PresetLimitExceededError") {
		t.Error("presets.controller.ts must handle PresetLimitExceededError and map it to 422 status")
	}
}

func TestPresetsControllerMapsPresetLimitErrorToCode(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "PRESET_LIMIT_EXCEEDED") {
		t.Error("presets.controller.ts must return PRESET_LIMIT_EXCEEDED error code in response")
	}
}

func TestPresetsControllerUsesPresetService(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "presetService") {
		t.Error("presets.controller.ts must use the presetService singleton")
	}
}

func TestPresetsControllerUsesZodValidation(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	lower := strings.ToLower(ctrl)
	if !strings.Contains(lower, "zod") && !strings.Contains(lower, "safeParse") && !strings.Contains(lower, "z.object") {
		t.Error("presets.controller.ts must use zod validation for request bodies")
	}
}

func TestPresetsControllerValidatesNameRequired(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "name") {
		t.Error("presets.controller.ts zod schema must require the name field")
	}
}

func TestPresetsControllerValidatesNameMaxLength(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "100") {
		t.Error("presets.controller.ts zod schema must enforce max 100 chars on name")
	}
}

func TestPresetsControllerHandles403Or404ForOtherUser(t *testing.T) {
	ctrl := readPresetsControllerTS(t)
	if !strings.Contains(ctrl, "403") && !strings.Contains(ctrl, "404") {
		t.Error("presets.controller.ts must handle 403/404 errors for cross-user operations")
	}
}

// ---- presets.test.ts integration test content ----

func TestPresetsIntegrationTestMocksPresetService(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "jest.mock") {
		t.Error("presets.test.ts must mock the preset service before importing the app")
	}
}

func TestPresetsIntegrationTestImportsApp(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "app") {
		t.Error("presets.test.ts must import the app for supertest requests")
	}
}

func TestPresetsIntegrationTestContainsGetPresetsEndpoint(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	lower := strings.ToLower(testFile)
	if !strings.Contains(lower, "get /api/presets") && !strings.Contains(testFile, "get('/api/presets") {
		t.Error("presets.test.ts must contain tests for GET /api/presets")
	}
}

func TestPresetsIntegrationTestContainsPostPresetsEndpoint(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	lower := strings.ToLower(testFile)
	if !strings.Contains(lower, "post /api/presets") && !strings.Contains(testFile, "post('/api/presets") {
		t.Error("presets.test.ts must contain tests for POST /api/presets")
	}
}

func TestPresetsIntegrationTestContainsPutPresetsEndpoint(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	lower := strings.ToLower(testFile)
	if !strings.Contains(lower, "put /api/presets") && !strings.Contains(testFile, "put('/api/presets") {
		t.Error("presets.test.ts must contain tests for PUT /api/presets/:id")
	}
}

func TestPresetsIntegrationTestContainsDeletePresetsEndpoint(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	lower := strings.ToLower(testFile)
	if !strings.Contains(lower, "delete /api/presets") && !strings.Contains(testFile, "delete('/api/presets") {
		t.Error("presets.test.ts must contain tests for DELETE /api/presets/:id")
	}
}

func TestPresetsIntegrationTestVerifies200OnGetPresets(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "200") {
		t.Error("presets.test.ts must verify 200 status on GET /api/presets")
	}
}

func TestPresetsIntegrationTestVerifiesEmptyArrayForNewUser(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "empty") && !strings.Contains(testFile, "toHaveLength(0)") && !strings.Contains(testFile, "length).toBe(0)") {
		t.Error("presets.test.ts must verify GET /api/presets returns empty array for a new user")
	}
}

func TestPresetsIntegrationTestVerifies201OnCreate(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "201") {
		t.Error("presets.test.ts must verify 201 status on successful POST /api/presets")
	}
}

func TestPresetsIntegrationTestVerifies422OnLimitExceeded(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "422") {
		t.Error("presets.test.ts must verify 422 status when preset limit is exceeded")
	}
}

func TestPresetsIntegrationTestVerifiesPresetLimitExceededError(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "PRESET_LIMIT_EXCEEDED") {
		t.Error("presets.test.ts must verify the PRESET_LIMIT_EXCEEDED error code in 422 response")
	}
}

func TestPresetsIntegrationTestVerifies403Or404ForOtherUserUpdate(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "403") && !strings.Contains(testFile, "404") {
		t.Error("presets.test.ts must verify 403 or 404 when updating another user's preset")
	}
}

func TestPresetsIntegrationTestVerifies204OnDelete(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "204") {
		t.Error("presets.test.ts must verify 204 status on successful DELETE /api/presets/:id")
	}
}

func TestPresetsIntegrationTestVerifiesZodNameRequired(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "name") {
		t.Error("presets.test.ts must include zod validation tests for required name field")
	}
}

func TestPresetsIntegrationTestVerifiesZodNameMaxLength(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "100") {
		t.Error("presets.test.ts must include zod validation test for name max 100 characters")
	}
}

func TestPresetsIntegrationTestVerifiesPutReturns200(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	// The file should test PUT returning 200
	if !strings.Contains(testFile, "200") {
		t.Error("presets.test.ts must verify 200 status on successful PUT /api/presets/:id")
	}
}

func TestPresetsIntegrationTestContainsPresetIdParam(t *testing.T) {
	testFile := readPresetsIntegrationTestTS(t)
	if !strings.Contains(testFile, "/api/presets/") {
		t.Error("presets.test.ts must test single-preset endpoints using /api/presets/:id pattern")
	}
}
